src/expressions.js

```diff
@@ -40,3 +40,37 @@ exports.satisfiesExpression = function(scopeset, expr) {
     exports.satisfiesExpression(scopeset, subexpr)
   );
 };
+
+/**
+ * Given a scopeset and a scope-expression, remove any scopes
+ * that are in the scopeset so that all remaining scopes
+ * are ones that are missing. If all scopes are provided, this
+ * function will return null. This function is useful to
+ * generate nice error messages about which scopes a client
+ * is missing.
+ *
+ * This is separate from satisfiesExpression in order to keep
+ * that code path simple and clean due to its important nature.
+ *
+ * Returns a scope expression where all scopes that exist are
+ * missing from the scopeset. Any scopes under an AllOf key
+ * are definitely needed to satisfy the expression and at least
+ * one of the scopes under an AnyOf must be provided to satisfy.
+ */
+exports.removeGivenScopes = function(scopeset, expr) {
+  assert(Array.isArray(scopeset), 'Scopeset must be an array.');
+
+  if (typeof expr === 'string') {
+    if (scopeset.some(scope => patternMatch(scope, expr))) {
+      return null;
+    }
+    return expr;
+  }
+
+  const operator = Object.keys(expr)[0];
+  const remaining = expr[operator].map(e => exports.removeGivenScopes(scopeset, e)).filter(e => e !== null);
+  if (remaining.length === 0) {
```

Comment:
This is incorrect.

There are two cases:
1. AllOf which **after** filter is empty, which implies return `null`
2. AnyOf which **before** filter contains a `null` which implies return `null`